warp10/src/main/java/io/warp10/script/binary/MUL.java

```diff
@@ -177,13 +170,20 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
       if (!(gts.getType() == TYPE.LONG || gts.getType() == TYPE.DOUBLE)) {
         throw new WarpScriptException(typeCheckErrorMsg);
       }
-      
-      double op = op1gts ? ((Number) op2).doubleValue() : ((Number) op1).doubleValue();
-      
-      for (int i = 0; i < n; i++) {
-        double value = ((Number) GTSHelper.valueAtIndex(gts, i)).doubleValue() * op;
-        GTSHelper.setValue(result, GTSHelper.tickAtIndex(gts, i), GTSHelper.locationAtIndex(gts, i), GTSHelper.elevationAtIndex(gts, i), value, false);
-      }      
+
+      Number op = op1gts ? (Number) op2 : (Number) op1;
+
+      if (op instanceof Double || gts.getType() == TYPE.DOUBLE) {
+        for (int i = 0; i < n; i++) {
+          double value = ((Number) GTSHelper.valueAtIndex(gts, i)).doubleValue() * op.doubleValue();
+          GTSHelper.setValue(result, GTSHelper.tickAtIndex(gts, i), GTSHelper.locationAtIndex(gts, i), GTSHelper.elevationAtIndex(gts, i), value, false);
+        }
+      } else {
+        for (int i = 0; i < n; i++) {
+          long value = ((Number) GTSHelper.valueAtIndex(gts, i)).longValue() * op.longValue();
```

Comment:
avoid op.longValue()

Reply:
Ditto.

Reply:
GET THIS FREAKING STATEMENT OUT OF THE LOOP!!!! ;-)